base/slicearray.jl

```diff
@@ -0,0 +1,234 @@
+
+Create a [`Slices`](@ref) object that is an array of slices over dimensions `dims` of `A`, returning
+views that select all the data from the other dimensions in `A`. `dims` can either by an
+integer or a tuple of integers.
+
+If `drop = true` (the default), the outer `Slices` will drop the inner dimensions, and
+the ordering of the dimensions will match those in `dims`. If `drop = false`, then the
+`Slices` will have the same dimensionality as the underlying array, with inner
+dimensions having size 1.
+
+See also [`eachrow`](@ref), [`eachcol`](@ref), [`mapslices`](@ref) and [`selectdim`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.9"
+     Prior to Julia 1.9, this returned an iterator, and only a single dimension `dims` was supported.
+
+# Example
+
+```jldoctest
+julia> m = [1 2 3; 4 5 6; 7 8 9]
+3×3 Matrix{Int64}:
+ 1  2  3
+ 4  5  6
+ 7  8  9
+
+julia> s = eachslice(m, dims=1)
+3-element RowSlices{Matrix{Int64}, Tuple{Base.OneTo{Int64}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}:
+ [1, 2, 3]
+ [4, 5, 6]
+ [7, 8, 9]
+
+julia> s[1]
+3-element view(::Matrix{Int64}, 1, :) with eltype Int64:
+ 1
+ 2
+ 3
+
+julia> eachslice(M, dims=1, drop=false)
+3×1 Slices{Matrix{Int64}, Tuple{Int64, Colon}, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, 2}:
+ [1, 2, 3]
+ [4, 5, 6]
+ [7, 8, 9]
+```
+"""
+@inline function eachslice(A; dims, drop=true)
+    _eachslice(A, dims, drop)
+end
+
+"""
+    eachrow(A::AbstractVecOrMat) <: AbstractVector
+
+Create a [`RowSlices`](@ref) object that is a vector of rows of matrix or vector `A`.
+Row slices are returned as `AbstractVector` views of `A`.
+
+See also [`eachcol`](@ref), [`eachslice`](@ref) and [`mapslices`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.9"
+     Prior to Julia 1.9, this returned an iterator.
+
+# Example
+
+```jldoctest
+julia> a = [1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> s = eachrow(a)
+2-element RowSlices{Matrix{Int64}, Tuple{Base.OneTo{Int64}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}:
+ [1, 2]
+ [3, 4]
+
+julia> s[1]
+2-element view(::Matrix{Int64}, 1, :) with eltype Int64:
+ 1
+ 2
+```
+"""
+eachrow(A::AbstractMatrix) = _eachslice(A, (1,), true)
+eachrow(A::AbstractVector) = eachrow(reshape(A, size(A,1), 1))
+
+"""
+    eachcol(A::AbstractVecOrMat) <: AbstractVector
+
+Create a [`ColumnSlices`](@ref) object that is a vector of columns of matrix or vector `A`.
+Column slices are returned as `AbstractVector` views of `A`.
+
+See also [`eachrow`](@ref), [`eachslice`](@ref) and [`mapslices`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.7"
+     Prior to Julia 1.7, this returned an iterator.
```

Comment:
I guess this should refer to 1.9 as in the other docstrings?

```suggestion
!!! compat "Julia 1.9"
     Prior to Julia 1.9, this returned an iterator.
```